Feature/ajax2

Tekkon opened the pull request:
- Только автор может отредактировать свой вопрос
  - Только автор может отредактировать свой ответ
  - Переделать удаление ответов на ajax
  - Автор вопроса может выбрать лучший ответ для своего вопроса (лучший ответ может быть только 1)
  - Автор вопроса может выбрать другой ответ как лучший, если у вопроса уже выбран лучший ответ.
  - Если у вопроса выбран лучший ответ, то он отображается первым в списке ответов.

Commit:
Пользователь пытается отредактировать чужой ответ

Commit:
Только автор может отредактировать свой вопрос

Commit:
Переделал удаление ответов на ajax

Commit:
- Автор вопроса может выбрать лучший ответ для своего вопроса (лучший ответ может быть только 1)
 - Автор вопроса может выбрать другой ответ как лучший, если у вопроса уже выбран лучший ответ.

Commit:
Если у вопроса выбран лучший ответ, то он отображается первым в списке ответов.

darialapina commented on app/controllers/questions_controller.rb:
```diff
@@ -30,11 +30,7 @@ def create
-    else
-      render :edit
-    end
+    @question.update(question_params)
```
Нужно еще проверить, что не-автор не может изменить вопрос

darialapina commented on app/controllers/answers_controller.rb:
```diff
@@ -16,23 +16,27 @@ def destroy
 
   def update
     @answer = Answer.find(params[:id])
     @answer.update(answer_params)
```
Нужно еще проверить, что не-автор не может изменить ответ

darialapina commented on app/controllers/answers_controller.rb:
```diff
@@ -16,23 +16,27 @@ def destroy
     @answer.update(answer_params)
     @question = @answer.question
+
+    if @answer.favorite
```
Действия редактирования и выбора лучшего ответа лучще разделить на разные action'ы в контроллере.
Редактировать ответ может автор ответа.
Выбрать лучший - автор вопроса.
Если оставить в одном методе, то придётся проверять кто какие параметры меняет, лишняя сложность.

darialapina commented on app/views/answers/_answer.html.slim:
```diff
@@ -1,16 +1,25 @@
     p= answer.body
+    - if answer.favorite
+      p= 'Favorite!'
+    - if user_signed_in? && (current_user.author_of?(answer) || current_user.author_of?(answer.question))
```
В задании говорится:
**Только автор может отредактировать свой ответ**

vkurennov replied:
Да, у тебя получается, что и автор вопроса может редактировать любые ответы

darialapina commented on app/views/answers/_answer.html.slim:
```diff
@@ -1,16 +1,25 @@
       br
 
     p
       = form_for [@question, answer], remote: true, html: { id: "edit-answer-#{answer.id}" } do |f|
```
Форма с кнопкой сохранить будет показываться всем посетителям?

vkurennov replied:
Да, у тебя тег p в котором выводится форма (это кстати, несемантическая верстка) не вложен в if из 9-й строки.

darialapina commented on app/views/answers/update.js.erb:
```diff
@@ -1,8 +1,12 @@
+  $('#answer-<%= @answer.id %>-errors').html('<%= j render 'shared/error_messages', target: @answer %>');
+<% else %>
+  //$('#answer-<%= @answer.id %>').html('<%= j render @answer %>');
+  $('.answers').html('<%= j render @question.answers.sort_by { |a| a.favorite ? 0 : 1 } %>')
```
Обновлять весь список ответов - не хорошо, в прошлых заданиях такое поведение убирали.
Опять же, если разделить редактирование (вопрос останется на месте) и выбор лучшего (перенести див с ответом выше), то всё будет ок.

vkurennov replied:
И сортировку делать средствами руби - очень медленно и неэффективно, используй скоуп для сортировкуи в модели и примени его здесь, у тебя сразу будут отсортированные ответы

darialapina commented on app/views/questions/_question.html.slim:
```diff
@@ -0,0 +1,21 @@
+      br
+
+    p
+      = form_for question, remote: true, html: {id: "edit-question-#{question.id}"} do |f|
```
Форма (хоть и скрытая, судя по всему) будет рендериться для всех пользователей, нужно и её за проверку на автора убрать.

vkurennov replied:
Не исправлено, форма действительно не под условием, надо сдвинуть все вместе с тегом p под условие из 5-й строки

darialapina commented on app/views/questions/show.html.slim:
```diff
@@ -5,7 +5,7 @@ p= @question.body
 .answers
   h2 Answers
-  = render @question.answers
+  = render @question.answers.sort_by { |a| a.favorite ? 0 : 1 }
```
Ответы везде должны выводиться с лучшим наверху. Правильнее сделать default_scope в модели.

vkurennov replied:
Да, скоуп, но лучше не default_scope (у него есть определенные особенности, из-за которых не стоит его часто применять), а обычный именованый скоуп.

darialapina commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -49,24 +49,19 @@
-      it 'redirects to question view' do
-        delete :destroy, params: { question_id: question, id: answer.id }
-        expect(response).to redirect_to question_path(question)
+        expect { delete :destroy, params: { question_id: question, id: answer.id }, format: :js }.to change(Answer, :count).by(-1)
```
Тут можно проверить, что ответы уменьшились у конкретного вопроса, а не в целом в базе.

vkurennov replied:
На самом деле не критично :-)

darialapina commented on spec/controllers/answers_controller_spec.rb:
```diff
@@ -89,9 +84,32 @@
+
+      expect(answer.favorite).to eq true
+      expect(another_answer.favorite).to eq false
+    end
```
Нужны проверки на случай, когда не-автор пытается отредактировать ответ.

darialapina commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -95,37 +95,28 @@
-        patch :update, params: { id: question, question: attributes_for(:question) }
-        expect(response).to redirect_to question
-      end
+    it 'assigns requested question to @question' do
```
Тут аналогично ответу. Нет проверок на случай запроса от не-автора вопроса.
Это в целом к блоку `describe 'PATCH #update'`. Не могу прицепить комментарий в более правильное место)

darialapina commented on spec/features/create_answer_spec.rb:
```diff
@@ -13,7 +13,7 @@
     visit question_path question
 
-    fill_in 'Body', with: 'This is my answer.'
+    fill_in 'create-answer-body', with: 'This is my answer.'
```
Так тоже можно, но рассмотри возможность ограничить поиск элемента, например, по блоку с css-классом

> within '.answers' do
>   тут действия
> end

В будущем пригодится, id не напасёшься на все элементы)

darialapina replied:
Собственно, ты внизу эту конструкцию используешь, тебе она знакома)

Commit:
Изменить вопрос может только его автор

Commit:
только автор ответа может его изменить

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -16,23 +16,32 @@ def destroy
     end
   end
 
   def update
```
Метод должен выполнять одну задачу - либо обновлять, либо назначать ответ лучшим. У тебя в одном методе, фактически, объеденены 2.

vkurennov commented on app/controllers/answers_controller.rb:
```diff
@@ -16,23 +16,32 @@ def destroy
+      flash[:notice] = 'You can update only your answers.'
+    end
+
+    if @answer.favorite
```
Во-первых, для назначения ответа лучшим и для апдейта разные проверки на автора: назначить лучшим может только автор вопроса, а редактировать свой ответ - только автор ответа.

Во-вторых, нужно выносить всю бизнес-логику из контроллера в модели. Т.е. сделать метод в модели Answer, который делает ответ лучшим, а в котнроллере просто взывать его (также в отдельном экшене)

vkurennov commented on app/views/answers/_answer.html.slim:
```diff
@@ -1,16 +1,25 @@
+        - if user_signed_in? && current_user.author_of?(answer)
+          = f.label :body, "Answer"
+          = f.text_area :body
+        - if user_signed_in? && current_user.author_of?(answer.question)
```
Это не нужно делать через форму. Редактирование - это одна фича, назначение лучшего - вторая

Commit:
- Добавил экшен mark_favorite в контроллер AnswersController для выбора лучшего ответа
- В модели Answer и Question добавил методы для выбора лучшего ответа

Commit:
Добавил гем orderly для проверки порядка элементов на странице

Commit:
Добавил route на экшен mark_favorite

Commit:
Добавил скоуп для сортировки ответов (лучший ответ первый)

Commit:
- Убрал редактирование поля favorite из формы
- Добавил радиобаттоны для выбора лучшего ответа
- Радиобатанны выводим только для автора вопроса
- Ссылку на редактирование и форму редактироания ответа выводим только для автора ответа

Commit:
Вернул рендеринг только измененного ответа

Commit:
Выводим ответы отсортированными (лучший ответ - первый)

Commit:
Тесты для выбора лучшего ответа

Commit:
Тесты выбора лучшего ответа

Commit:
Переносим выбранный лучший ответ в начало списка

Commit:
Добавил проверки на автора и не-автора ответа

Commit:
Убрал id у поля body, в спеке находим поле по названию

darialapina commented on app/models/answer.rb:
```diff
@@ -2,5 +2,12 @@ class Answer < ApplicationRecord
+  scope :ordered_by_favorite, -> { order(favorite: :desc) }
+
+  def mark_favorite
+    self.update(favorite: true)
```
Тут стоит применить [транзакцию](http://api.rubyonrails.org/classes/ActiveRecord/Transactions/ClassMethods.html), чтобы изменения были атомарными.

darialapina commented on spec/controllers/questions_controller_spec.rb:
```diff
@@ -114,6 +117,13 @@
+      patch :update, id: another_question, question: { title: 'new title', body: 'new body' }, format: :js
+      question.reload
+      expect(question.title).to include 'What happened write after the big bang?'
+      expect(question.body).to include 'I really want to know!'
```
Ты в запросе пытаешься изменить another_question, а потом проверяешь question
Поэтому тест зеленый.
Получилось, что у ты в этом тесте залогинен под автором another_question ( он же автор question)

Commit:
Поправил тест.

Commit:
Применил транзакцию, чтобы изменения были атомарными

Commit:
Разбил спеки в features по папками

Commit:
Добавил хелпер, который ожидает завершения выполнения всех ajax-запросов

vkurennov commented on app/models/answer.rb:
```diff
@@ -2,5 +2,14 @@ class Answer < ApplicationRecord
+  def mark_favorite
+    ActiveRecord::Base.transaction do
+      self.update!(favorite: true)
+      self.question.unmark_favorites(self)
```
Вот это не стоило выносить в метод, можно было бы все прямо здесь сделать. Ну и нужно наоборот сделать - сначала убрать флаг лучшего ответа с других, а затем - сделать `update!` для текущего ответа

vkurennov commented on app/views/answers/_answer.html.slim:
```diff
@@ -1,16 +1,23 @@
+    - if user_signed_in? && current_user.author_of?(@question)
+      = label :favorite, 'Favorite'
+      = radio_button 'rb', 'favorite', "answer_#{answer.id}", checked: answer.favorite, class: 'favorite-check', data: { remote: true, method: :put,
+              url: "/questions/#{@question.id}/answers/#{answer.id}/mark-favorite" }
```
Так, а почему url-хелперы не используешь? Посмотри на rake routes и используй нужный метод.

vkurennov commented on config/routes.rb:
```diff
@@ -3,6 +3,8 @@
   resources :questions do
-    resources :answers
+    resources :answers do
+      put '/mark-favorite' => 'answers#mark_favorite'
```
Кастомные экшены в ресурсах принято писать так:

`put :mark_favorite, on: :member`

Твой синтаксис используется только если задается роут вне ресурса.

vkurennov commented on spec/models/answer_spec.rb:
```diff
@@ -4,4 +4,17 @@
+      answer.mark_favorite
+      answer.reload
+
+      expect(answer.favorite).to eq true
```
Нужно еще проверить, что другие лучшие ответы стали обычными. Отдельным тестом. Проверить именно то, что это происходит при вызове mark_favorite

Commit:
- Всю логику перенес в один метод mark_favorite у answer
- Сначала убираем флаг favorite с других ответов, затем присваиваем флаг favorite текущему ответу

Commit:
- Добавил тест на проверку, что при выборе лучшего ответа другие лучшие ответы становятся обычными
- Убрал тест метода unmark_favorites из question_spec, т.к. удалил этот метод

Commit:
- Использовал url-хелпер для экшена mark_favorite
- Исправил route для экшена mark_favorite

Commit:
Форма была не под условием, исправил

Commit:
После изменения раута answer_id поменялось на id
